Enable gocritic linting, fix errors

dfarrell07 opened the pull request:
Disable if-else chain linting as it seems to only flag false positives
in our code base. Our two violations don't switch on a single condition,
so while one of them can be written as a switch, it's less clear.

Signed-off-by: Daniel Farrell <dfarrell@redhat.com>

dfarrell07 commented:
I'm not sure what we should do to fix this one:

```
main.go:107:18: appendAssign: append result not assigned to the same slice (gocritic)
		localSubnets = append(submSpec.ServiceCidr, submSpec.ClusterCidr...)
```

skitt commented:
> I'm not sure what we should do to fix this one:
> 
> ```
> main.go:107:18: appendAssign: append result not assigned to the same slice (gocritic)
> 		localSubnets = append(submSpec.ServiceCidr, submSpec.ClusterCidr...)
> ```

I think this should work:

```
		localSubnets = append(localSubnets, submSpec.ServiceCidr...)
		localSubnets = append(localSubnets, submSpec.ClusterCidr...)
```

dfarrell07 commented:
Thanks @skitt! Enabling other parts of gocritic, I'm now unsure what  (if anything) we should do about these:

```
pkg/util/util.go:101:68: wrapperFunc: use strings.ReplaceAll method in `strings.Replace(privateIP, ".", "-", -1)` (gocritic)
			CableName:     fmt.Sprintf("submariner-cable-%s-%s", clusterID, strings.Replace(privateIP, ".", "-", -1)),
			                                                                ^
pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
			if j == 0 && j == len(connections)-1 {
			   ^
pkg/cableengine/fake/cableengine.go:61:2: unnecessaryDefer: defer e.Unlock() is placed just before return (gocritic)
	defer e.Unlock()
	^
pkg/controllers/datastoresyncer/datastoresyncer.go:85:3: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
		if !util.CompareEndpointSpec(endpoints[i].Spec, d.localEndpoint.Spec) {
		^
test/e2e/framework/framework.go:226:1: unnamedResult: consider giving a name to these results (gocritic)
func (f *Framework) GetGatewayInformer(cluster framework.ClusterIndex) (cache.SharedIndexInformer, chan struct{}) {
^
test/e2e/framework/submariner_resources.go:23:1: unnamedResult: consider giving a name to these results (gocritic)
func NoopCheckEndpoint(endpoint *submarinerv1.Endpoint) (bool, string, error) {
^
```

skitt commented:
> ```
> pkg/util/util.go:101:68: wrapperFunc: use strings.ReplaceAll method in `strings.Replace(privateIP, ".", "-", -1)` (gocritic)
> 			CableName:     fmt.Sprintf("submariner-cable-%s-%s", clusterID, strings.Replace(privateIP, ".", "-", -1)),
> 			                                                                ^
> ```

That one seems sensible, doesn’t `strings.ReplaceAll()` work here?

> ```
> pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
> 			if j == 0 && j == len(connections)-1 {
> 			   ^
> ```

Hah, that one caused comments in the PR review, I’d have to revisit it to see...

> ```
> pkg/cableengine/fake/cableengine.go:61:2: unnecessaryDefer: defer e.Unlock() is placed just before return (gocritic)
> 	defer e.Unlock()
> 	^
> ```

Not sure about this one; I’m not sure we’d remember to restore the `defer` if we ever add code between the unlock and the `return`. Perhaps `gocritic` flags that too...

> ```
> pkg/controllers/datastoresyncer/datastoresyncer.go:85:3: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
> 		if !util.CompareEndpointSpec(endpoints[i].Spec, d.localEndpoint.Spec) {
> 		^
> ```

What’s it like if you follow the directions?

> ```
> test/e2e/framework/framework.go:226:1: unnamedResult: consider giving a name to these results (gocritic)
> func (f *Framework) GetGatewayInformer(cluster framework.ClusterIndex) (cache.SharedIndexInformer, chan struct{}) {
> ^
> test/e2e/framework/submariner_resources.go:23:1: unnamedResult: consider giving a name to these results (gocritic)
> func NoopCheckEndpoint(endpoint *submarinerv1.Endpoint) (bool, string, error) {
> ^
> ```

This one is marked experimental, not sure it’s worth enabling now.

dfarrell07 commented:
Thanks again @skitt! I fixed the obvious-in-retrospect ones and added ignored the unnecessaryDefer and unnamedResult rules. Left the libreswan.go:323 failure for now. That logic is also flagged by the ifElseChain rule I have disabled FYI.

skitt commented on .golangci.yml:
```diff
@@ -1,4 +1,15 @@
+  gocritic:
+    enabled-tags:
+      - diagnostic
+      - experimental
```
Do we want to rely on experimental gocritic components in general? Some of them might be worth enabling piece-meal, but I’m not sure enabling the whole tag is a good idea.

dfarrell07 replied:
I'm not sure, good question. I think the experimental tag caught a lot of real things to improve relative to the few false positives/rules we don't like. Running with only the experimental tag against master, I get:

```
pkg/util/util.go:122:2: commentFormatting: put a space between `//` and comment text (gocritic)
	// length is 11
	^
pkg/util/util.go:241:9: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
	} else {
	       ^
pkg/util/util.go:242:6: sloppyReassign: re-assignment to `err` can be replaced with `err := ipt.Insert(table, chain, 1, ruleSpec...)` (gocritic)
		if err = ipt.Insert(table, chain, 1, ruleSpec...); err != nil {
		   ^
pkg/util/util.go:101:68: wrapperFunc: use strings.ReplaceAll method in `strings.Replace(privateIP, ".", "-", -1)` (gocritic)
			CableName:     fmt.Sprintf("submariner-cable-%s-%s", clusterID, strings.Replace(privateIP, ".", "-", -1)),
			                                                                ^
pkg/cable/strongswan/strongswan.go:418:5: sloppyReassign: re-assignment to `err` can be replaced with `err := i.renderCharonConfigTemplate(f)` (gocritic)
	if err = i.renderCharonConfigTemplate(f); err != nil {
	   ^
pkg/cable/wireguard/driver.go:44:2: commentFormatting: put a space between `//` and comment text (gocritic)
	//TODO use submariner prefix
	^
pkg/cable/wireguard/driver.go:230:27: commentFormatting: put a space between `//` and comment text (gocritic)
			Port: w.spec.NATTPort, //TODO move port to endpoint spec
			                       ^
pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
			if j == 0 && j == len(connections)-1 {
			   ^
pkg/cableengine/cableengine.go:83:5: sloppyReassign: re-assignment to `err` can be replaced with `err := i.driver.Init()` (gocritic)
	if err = i.driver.Init(); err != nil {
	   ^
pkg/cableengine/fake/cableengine.go:61:2: unnecessaryDefer: defer e.Unlock() is placed just before return (gocritic)
	defer e.Unlock()
	^
pkg/cableengine/syncer/syncer.go:62:2: ifElseChain: rewrite if-else to switch statement (gocritic)
	if errors.IsNotFound(err) {
	^
pkg/controllers/datastoresyncer/datastoresyncer.go:85:3: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
		if !util.CompareEndpointSpec(endpoints[i].Spec, d.localEndpoint.Spec) {
		^
pkg/routeagent/controllers/route/route.go:138:38: captLocal: `ClusterCidr' should not be capitalized (gocritic)
func NewController(clusterID string, ClusterCidr []string, ServiceCidr []string, objectNamespace string,
                                     ^
pkg/routeagent/controllers/route/route.go:799:10: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
		} else {
		       ^
pkg/routeagent/controllers/route/route.go:901:11: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
			} else {
			       ^
pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go:15:2: dupImport: package is imported 2 times under different aliases on lines 15 and 16 (gocritic)
	clientsetv1 "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
	^
pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go:16:2: dupImport: package is imported 2 times under different aliases on lines 15 and 16 (gocritic)
	submarinerClientset "github.com/submariner-io/submariner/pkg/client/clientset/versioned/typed/submariner.io/v1"
	^
test/e2e/framework/framework.go:226:1: unnamedResult: consider giving a name to these results (gocritic)
func (f *Framework) GetGatewayInformer(cluster framework.ClusterIndex) (cache.SharedIndexInformer, chan struct{}) {
^
test/e2e/framework/submariner_resources.go:23:1: unnamedResult: consider giving a name to these results (gocritic)
func NoopCheckEndpoint(endpoint *submarinerv1.Endpoint) (bool, string, error) {
^
main.go:107:18: appendAssign: append result not assigned to the same slice (gocritic)
		localSubnets = append(submSpec.ServiceCidr, submSpec.ClusterCidr...)
		               ^
main.go:110:5: emptyStringTest: replace `len(submSpec.CableDriver) == 0` with `submSpec.CableDriver == ""` (gocritic)
	if len(submSpec.CableDriver) == 0 {
	   ^
pkg/globalnet/controllers/ipam/globalnet.go:9:2: dupImport: package is imported 2 times under different aliases on lines 9 and 10 (gocritic)
	k8sv1 "k8s.io/api/core/v1"
	^
pkg/globalnet/controllers/ipam/globalnet.go:10:2: dupImport: package is imported 2 times under different aliases on lines 9 and 10 (gocritic)
	v1 "k8s.io/api/core/v1"
	^
```

dfarrell07 replied:
Removed the experimental tag in the latest. :+1:

tpantelis commented:
> > ```
> > pkg/cableengine/fake/cableengine.go:61:2: unnecessaryDefer: defer e.Unlock() is placed just before return (gocritic)
> > 	defer e.Unlock()
> > 	^
> > ```
> 
> Not sure about this one; I’m not sure we’d remember to restore the `defer` if we ever add code between the unlock and the `return`. Perhaps `gocritic` flags that too...
> 

The point here is to synchronize access to the `HAStatus` field for memory consistency, ie perform the read while holding the lock.  It seems `gocritic` doesn't understand that. We can also do it this way which should avoid the error:

```
e.Lock()
ret := e.HAStatus
e.Unlock()
return ret
```

> > ```
> > pkg/controllers/datastoresyncer/datastoresyncer.go:85:3: nestingReduce: invert if cond, replace body with `continue`, move old body after the statement (gocritic)
> > 		if !util.CompareEndpointSpec(endpoints[i].Spec, d.localEndpoint.Spec) {
> > 		^
> > ```
> 
> What’s it like if you follow the directions?

It's a style preference - their suggestion reduces the amount of indented code which is fine. 

> 
> > ```
> > test/e2e/framework/framework.go:226:1: unnamedResult: consider giving a name to these results (gocritic)
> > func (f *Framework) GetGatewayInformer(cluster framework.ClusterIndex) (cache.SharedIndexInformer, chan struct{}) {
> > ^
> > test/e2e/framework/submariner_resources.go:23:1: unnamedResult: consider giving a name to these results (gocritic)
> > func NoopCheckEndpoint(endpoint *submarinerv1.Endpoint) (bool, string, error) {
> > ^
> > ```
> 
> This one is marked experimental, not sure it’s worth enabling now.

yeah there's pros and cons to named result params - mainly a style thing for documentation to clarify meaning when there's multiple parameters of the same type. I think it's flagging these functions b/c they return multiple values other than `error`. I don't think it's worth failing linting.

tpantelis commented on pkg/globalnet/controllers/ipam/ipam.go:
```diff
@@ -210,10 +210,10 @@ func (i *Controller) processNextObject(objWorkqueue workqueue.RateLimitingInterf
 }
 
-func (i *Controller) enqueueObject(obj interface{}, workqueue workqueue.RateLimitingInterface) {
+func (i *Controller) enqueueObject(obj interface{}, rateLim workqueue.RateLimitingInterface) {
```
Why does it complain about the name `workqueue`? I prefer that as that's what it really is with rate limiting characteristic. Or rename to `queue` if that satisfies it.

dfarrell07 replied:
Because it shaddows the imported `k8s.io/client-go/util/workqueue`.

dfarrell07 replied:
Changed it to `queue` in the latest, thanks. :+1:

tpantelis commented on pkg/globalnet/controllers/ipam/ipam.go:
```diff
@@ -634,7 +634,7 @@ func (i *Controller) nodeUpdater(obj runtime.Object, key string) error {
-func logAndRequeue(key string, workqueue workqueue.RateLimitingInterface) {
-	klog.V(log.DEBUG).Infof("%s enqueued %d times", key, workqueue.NumRequeues(key))
-	workqueue.AddRateLimited(key)
+func logAndRequeue(key string, rateLim workqueue.RateLimitingInterface) {
```
Same as above.

dfarrell07 commented:
> e.Lock()
> ret := e.HAStatus
> e.Unlock()
> return ret

Changed it to this and removed the ignore in the latest.

dfarrell07 commented:
Thanks for the great reviews @tpantelis and @skitt! The last still flagged is:

```
pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
			if j == 0 && j == len(connections)-1 {
			   ^
```

skitt commented:
> ```
> pkg/cable/libreswan/libreswan.go:323:7: badCond: `j == 0 && j == len(connections)-1` condition is suspicious (gocritic)
> 			if j == 0 && j == len(connections)-1 {
> 			   ^
> ```

https://github.com/submariner-io/submariner/pull/715

tpantelis commented on pkg/cable/strongswan/strongswan.go:
```diff
@@ -415,7 +415,7 @@ func (i *strongSwan) writeCharonConfig(path string) error {
 	}
 
-	if err = i.renderCharonConfigTemplate(f); err != nil {
+	if err := i.renderCharonConfigTemplate(f); err != nil {
```
It's interesting that it complains about "sloppy reassign" here but not on L422 which does the same thing.

dfarrell07 replied:
Good point, I don't see why that be the case.

tpantelis commented:
More jobs failing with: 

```
Step 1/6 : FROM registry.fedoraproject.org/fedora-minimal:32
manifest for registry.fedoraproject.org/fedora-minimal:32 not found: manifest unknown: manifest unknown
```
